### PR TITLE
Slim version of docker-compose for OSS including Nginx

### PR DESCRIPTION
**Pull request by uCatu**
Connected to #2 
#### What does this PR do?
This PR:
- Integrating Nginx to docker-compose OSS (allowing TLS communication).
- Allowing use of selfsigned certificates using openssl image.
- Defaulting to use of 3rd party certificates.

Elabroting in `README` how to use 3rd party & selfsigend certificate, this section can be (should be) to online docs @shuligh 

Using Ruby/Rails is **not** the selected approach:
Screen Shot 2019-10-24 at 13 56 00
@shaharglazner 

#### Where should the reviewer start?
- Is it making our customer life easy?
- Is it secure enough?
- What are the alternatives?

####  Next steps:
~~Use openssl that is already inside Conjur image
We can look at master-key generation process where we quickly spinning up conjur container and then deleting it. Saving us the need for open-ssl image.~~

> Harden and best practice configuration of Nginx container
If we have guidelines for user adding a reverse-proxy, we should also follow it in our Nginx container  (± 1 day, if needed at all)

~~Document the process of supplying own certificates
Document replacing certificate in TLS mounted folder, check that it works. (±3 hours)~~

### CODE REVIEW

**Comment by jvanderhoof** [comment]
@uCatu: It looks like a great start.  It looks like we're generating a custom certificate on startup (using the `openssl` container.  Would it make more sense to document the process of generating those certificates (or wrap it up into a simple bash script)? Anyone concerned with TLS is likely to be bringing their own certificates, considering this example will display the untrusted host warning when viewed with a browser.

**Comment by jvanderhoof** [comment]
@uCatu, @boazmichaely, @garkler: IMO, we can merge this PR (it's not a clear step in the wrong direction), but this PR does not move us toward our goal of Conjur OSS with TLS in any meaningful way.  Looking at the acceptance criteria:

> Is it making our customer life easy?

Not really as it only generates a self signed certificate.  It does not include directions for accessing the public signing certificate (which a user would need to establish a secure connection), nor does it include any instructions for configuring your own certificate.

> Is it secure enough?

IMO, no, as it generates a self signed certificate, which users need to trust without being able to verify.  Including instructions for accessing the public signing certificate or adding your own certificate would be adequate.

> What are the alternatives?

IMO, adding some documentation to the Readme with an example of how to load your own certificates would be adequate toward resolving all these concerns.

**Comment by boazmichaely** [comment]
@jvanderhoof thanks for pointing out the gaps.
My understanding from this is that the only missing piece is documentation?

**Comment by uCatu** [comment]
@jvanderhoof @boazmichaely 
How should we approache this PR now that we have new (decompose) architacture of OSS including Ngnix?

**Comment by boazmichaely** [comment]
Hi,
Making sure we are closed on this issue. This has nothing to do with the decomposed architecture for DAP…
Thanks,
Boaz

File: conf/certificate/.gitignore [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Ignore everything in this directory
```
Comment (jvanderhoof) [inline conf/certificate/.gitignore]: Do we need this file?  Shouldn't the `.gitignore` in the root directory handle this?

File: init.sh [hunk 2]
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env bash
+
+# Clean previous run data.
+rm -f admin_data data_key
+
+docker-compose stop
+if [ $? -eq 0 ]; then
```
Comment (jvanderhoof) [inline init.sh]: This logic shows up a lot in this script.  What about abstracting it out into a method:
```sh
function execute() {
  eval $1
  if [ $? -eq 0 ]; then
    echo "$1... OK"
  else
    echo "$1... FAIL"
  fi
}
```
Which will allow us to call these methods with:
```sh
execute 'docker-compose stop'
```

File: conf/certificate/.gitignore (hunk 1 above)
Comment (uCatu) [reply conf/certificate/.gitignore]: Yes, we need this **empty** directory to be created
https://git.wiki.kernel.org/index.php/GitFaq#Can_I_add_empty_directories.3F

File: init.sh (hunk 2 above)
Comment (uCatu) [reply init.sh]: Oops it was meant to be only for dev, but I think it can be useful - I added that function and completed all 4 Units

**Comment by uCatu** [comment]
> I'd love to see some documentation about how a user could configure Conjur with custom certificates. Most customers would only use generated certs for testing.

@jvanderhoof 
I tried as much as possible to eleborate how to add custom certificate in our `readme`
https://github.com/cyberark/conjur-quickstart/blob/oss/README.md

Maybe I think that I might be missing something because I don't know what else to share with customer to supply their own certificate - they need to generate it by themself (https://letsencrypt.org/) or get it from their IT... and follow readme instructions...

**Comment by uCatu** [comment]
> If a customer is comfortable with an invalid certificate and chooses to run Conjur without a load balancer, this example helps highlight the necessary setup. I'm OK merging this work.
> 
> Moving forward, we'll plan on documenting more realistic production grade setups.

@jvanderhoof Sorry I don't want to merge this PR if you don't see any use to it, it's not a rubber stamp solution. 

I feel like I'm repeating myself for the last copule of months. Can you pelease eleborate what you see is missing here from making this solution a production grade setup (keeping in mind it's a "quick-start" setup).

We have **defaulted Nginx to customer supplied certificate**, we don't issue a self-signed.
ONLY if customer explicitly run a command that request openssl to generate one.

SEE README:

> To run Quickstart in an unsecured manner, using a self signed certificate, replace command:
> 
> <code>docker-compose up -d</code>
> 
> With following line:
> 
> <code>docker-compose -f docker-compose.yml -f docker-compose.selfsigned.yml up -d</code>


@boazmichaely @InbalZilberman @hilagross @rrefael

**Comment by jvanderhoof** [comment]
Honestly, I just don't have the headspace to focus on this right now.  If we want to merge this, I'm in full support.

**Comment by sgnn7** [comment]
@uCatu Can you update this PR? Currently it cannot be merged in as-is and the underlying repo changed to include some of your changes since you opened your PR.

**Comment by sgnn7** [comment]
It's been a year since this PR has been opened and it's behind master so I'm closing it. @uCatu re-open a new one when needed.

**Comment by boazmichaely** [comment]
@sgnn7 @izgeri let's review how this can get into the master instead of closing it. I've been the one to hold it back because I had not completed the review but I'd like us to review options

**Comment by sgnn7** [comment]
@boazmichaely It's fine to have a PR closed until it's ready to review - I didn't delete the original branch so it can still be worked on. 

As far as I understand the situation:
- We need a team/person with bandwidth to work on this since it requires a medium chunk of rework (the repo has had updates since this PR was open and it can't be merged as-is).
- Leaving this PR open in a broken state is of no value to the team as it can be reopened at any point but as-is it keeps pinging us in chat daily as a PR that we haven't merged this yet (and we can't).
- Some clear guidance on what is further needed from you in terms of bigger picture is needed before even that rework mentioned earlier can be done otherwise the rework could be wasted work.

**Comment by izgeri** [comment]
@boazmichaely I agree with Srdjan - I think if we want this work to continue, we should close the PR and open an issue in this repo that refers to the desired changes and links to the WIP branch.

I am going to close this PR for now, and I would be happy to create the issue - but looking at the diff versus the PR description I am not sure I could clearly describe what the main benefit of this chunk of work is in a new issue - taking the convo to Slack.

**Pull request closed without merging**